src/plugins/direct_access_links/public/direct_access_link_generator.ts

```diff
@@ -0,0 +1,174 @@
+  [key: string]: GeneratorState<any, string | undefined, object | undefined>;
+}
+
+export interface AccessLinkGenerator<Id extends GeneratorId> {
+  id: Id;
+  createUrl(state: GeneratorStateMapping[Id]['State']): Promise<string>;
+  isDeprecated: boolean;
+  migrate(
+    state: GeneratorStateMapping[Id]['State']
+  ): Promise<{
+    state: GeneratorStateMapping[Id]['MigratedState'];
+    id: GeneratorStateMapping[Id]['MigratedId'];
+  }>;
+}
+
+const noMigrationFnProvidedWarningText = i18n.translate(
+  'directAccessLinks.error.noMigrationFnProvided',
+  {
+    defaultMessage:
+      'If the access link generator is marked as deprecated, you must provide a migration function.',
+  }
+);
+
+export interface DirectAccessLinkOptions<Id extends GeneratorId> {
+  id: Id;
+  createUrl?: (state: GeneratorStateMapping[Id]['State']) => Promise<string>;
+  isDeprecated?: boolean;
+  migrate?: (
+    state: GeneratorStateMapping[Id]['State']
+  ) => Promise<{
+    state: GeneratorStateMapping[Id]['MigratedState'];
+    id: GeneratorStateMapping[Id]['MigratedId'];
+  }>;
+}
+
+export const createDirectAccessLinkGenerator = <Id extends GeneratorId>(
+  options: DirectAccessLinkOptions<Id>,
+  getGenerator: DirectAccessLinksStart['getAccessLinkGenerator']
+): AccessLinkGenerator<Id> => {
+  if (options.isDeprecated && !options.migrate) {
+    throw new Error(noMigrationFnProvidedWarningText);
+  }
+
+  if (!options.isDeprecated && options.migrate) {
+    throw new Error(
+      i18n.translate('directAccessLinks.error.migrationFnGivenNotDeprecated', {
+        defaultMessage:
+          'If you provide a migration function, you must mark this generator as deprecated',
+      })
+    );
+  }
+
+  if (!options.createUrl && !options.isDeprecated) {
+    throw new Error(
+      i18n.translate('directAccessLinks.error.noCreateUrlFnProvided', {
+        defaultMessage:
+          'This generator is not marked as deprecated. Please provide a createUrl fn.',
+      })
+    );
+  }
+
+  if (options.createUrl && options.isDeprecated) {
+    throw new Error(
+      i18n.translate('directAccessLinks.error.createUrlFnProvided', {
+        defaultMessage: 'This generator is marked as deprecated. Do not supply a createUrl fn.',
+      })
+    );
+  }
+
+  return {
+    id: options.id,
+    createUrl: async (state: GeneratorStateMapping[Id]['State']) => {
+      if (options.createUrl && options.isDeprecated) {
+        throw new Error(
+          i18n.translate('directAccessLinks.error.createUrlFnProvided', {
+            defaultMessage: 'This generator is marked as deprecated. Do not supply a createUrl fn.',
+          })
+        );
+      }
+
+      if (!options.createUrl && !options.isDeprecated) {
+        throw new Error(
+          i18n.translate('directAccessLinks.error.noCreateUrlFnProvided', {
+            defaultMessage:
+              'This generator is not marked as deprecated. Please provide a createUrl fn.',
+          })
+        );
+      }
```

Comment:
I guess these checks are already available on lines 93&ndash;108? Should we delete these?

Reply:
yea, they do take up a lot of space. I added them because they help with the typescript checks. I can get rid of them but need to use `!` more.